docs/en/reference/testing.rst

```diff
@@ -0,0 +1,74 @@
+2. Make sure that the problem being solved needs to be solved in the DBAL.
+3. Document the proper usage of the DBAL APIs.
+
+Requirements
+------------
+
+1. Each pull request that adds new logic or changes the existing logic must have tests.
+2. The test that covers certain logic must fail without the changes in the code it covers.
+
+Unit Tests
+----------
+
+Unit tests are meant to cover the logic of a given unit (e.g. a class or a method) including the logic of its interaction
+with other units. In this case, the other units could be mocked.
+
+Unit tests are mostly welcomed for testing the logic that the DBAL itself defines (e.g. logging). They also could be
+used as a secondary tier of testing for the logic that requires integration testing.
+
+Integration Tests
+-----------------
+
+Integration (a.k.a. functional) tests are required when the behavior under test is dictated by the logic
+defined outside of the DBAL. It could be:
+
+1. The underlying database platform.
+2. The underlying database driver.
+3. SQL syntax and the standard as such.
+
+It is important to have integration tests for the cases above. Unlike unit tests, they make the external components
+the source of truth and help make sure that the logic implemented in the DBAL is correct even if the external components
+change (e.g. a new version of a database platform is supported).
+
+When Integration Testing Is Not Required
+----------------------------------------
+
+Some cases cannot be reproduced with the existing integration testing suite. It could be the scenarios that involve
+multiple concurrent database connections, transactions, locking, performance-related issues, etc.
+
+In such cases, it is still important that a pull request fixing the issues is accompanied with a free-form test that
+reproduces the issue being fixed.
+
+Testing Different Database Platforms
+------------------------------------
+
+Despite the fact that most of the issues are originally discovered on a specific database platform,
+the integration tests for all issues should be implemented by default at the database abstraction level
+and run against all the platforms that support the API being tested.
+
+This allows to ensure that the same scenario that was found failing on one platform also works on others. Or otherwise,
+the same issue could be reproduced on the platforms where it wasn't originally tested.
+
+If the newly added test fails on other platforms, and fixing it is out of the scope, the test can be explicitly marked
+as incomplete which will clearly identify the issue.
+
+Examples of such tests could be found under the the
+`Doctrine\Tests\DBAL\Functional\Platform <https://github.com/doctrine/dbal/tree/2.11.3/tests/Doctrine/Tests/DBAL/Functional/Platform>`_
+namespace.
+
+Other Notes
+-----------
+
+Do not mix the unit and the integration approaches in one test. Each of the approaches has its own area of application
```

Comment:
This needs the paths and separation strategies shown or explained, that are used in the repo.

Reply:
Could you elaborate on what you mean? I can actually add the details about how exactly the functional tests are written: the namespace, the base class and the auto-instantiated connection, something about paying more attention to the cleanup/teardown, but it would all go to the integration testing section.

As for the unit testing, there's not much to add.

Could you ask me some questions that I may try to answer?

Reply:
Yes, I meant the namespace, the base class and that stuff. I hadn't the cleanup in mind, but if you think this should be added too, then feel free to add it. `Doctrine\Tests\DBAL\Functional` is kind of hidden inside the namespaces/directories, so questions could be:

* Where do I put integration tests?
* Where do I put/setup fixtures for the integration tests? (Or: where does the data for integration tests come from?)

Because we already have examples in the project itself, there can be a link to one of those, so it doesn't have to be a detailed explanation.

For some people testing is something new and a link to the PHPUnit docs could be added too.